Source/ARTWebSocketTransport.m

```diff
@@ -149,28 +149,13 @@ - (NSURL *)setupWebSocket:(NSDictionary<NSString *, NSURLQueryItem *> *)params w
     // Format: MsgPack, JSON
     [queryItems addValueAsURLQueryItem:[_encoder formatAsString] forKey:@"format"];
 
-    if (options.recover) {
-        NSArray *recoverParts = [options.recover componentsSeparatedByString:@":"];
-        if (recoverParts.count > 1 && recoverParts.count <= 3) {
-            NSString *key = [recoverParts objectAtIndex:0];
-            NSString *serial = [recoverParts objectAtIndex:1];
-            ARTLogInfo(self.logger, @"R:%p WS:%p ARTWebSocketTransport: attempting recovery of connection %@", _delegate, self, key);
-
-            [queryItems addValueAsURLQueryItem:key forKey:@"recover"];
-            [queryItems addValueAsURLQueryItem:serial forKey:@"connectionSerial"];
-
-            int64_t msgSerial = [[recoverParts lastObject] longLongValue];
-            if (msgSerial) {
-                [_delegate realtimeTransportSetMsgSerial:self msgSerial:msgSerial];
-            }
-        }
-        else {
-            ARTLogError(self.logger, @"R:%p WS:%p ARTWebSocketTransport: recovery string is malformed, ignoring: '%@'", _delegate, self, options.recover);
-        }
+    // RTN16k
+    if (options.recover != nil) {
+        ARTConnectionRecoveryKey *const recoveryKey = [ARTConnectionRecoveryKey fromJsonString:options.recover];
```

Comment:
We decided to use exceptions here (not used anywhere else in the codebase, as it's not idiomatic for objective-c), because this is something that should never happen, as we serialize and de-serialize this JSON here inside this library (so it's never unpredictable data from the 3d party). Not sure we should handle this exception. WDYT @lawrence-forooghian

Reply:
> as we serialize and de-serialize this JSON here inside this library (so it's never unpredictable data from the 3d party)

There's nothing to stop the user from providing garbage data for the `recover` property, though; it accepts an arbitrary string.

Have either of you seen a spec point which describes what to do in the case of an error deserializing `options.recover`? If not, we probably should add one. How do we handle it in the SDKs you've worked on, @sacOO7? ably-js is pretty lax about it; if JSON deserialization fails [then it just continues as if the `recover` option were not provided](https://github.com/ably/ably-js/blob/fdd5a566eafac0a7fae9592f890361a608794487/src/common/lib/transport/connectionmanager.ts#L92-L98).

Reply:
>Have either of you seen a spec point which describes what to do in the case of an error deserializing options.recover (I can't see one)?

Me neither. I now log error and ignore invalid value like js SDK does - https://github.com/ably/ably-cocoa/pull/1714/commits/2df983784784d9d95268d3c2e12d01727faca9e4

Reply:
This one? @sacOO7